---
title: "Download ovarian cancer dataset from TCGA"
output: html_document
---

In this tutorial, we are going to download a dataset from The Cancer Genome Atlas. 
Please note that datasets are often huge and it will take some time.

First, letâ€™s focus on the TCGA datasets at [https://portal.gdc.cancer.gov](https://portal.gdc.cancer.gov/) 
and pick your favorite one.

For this tutorial, we choose TCGA-OV, i.e., High Grade Serous Ovarian Cancer (HGSOC) dataset.

Now, we'll set some variables and load some useful libraries. 
Moreover, to better organize downloads we suggest to create a directory ("downloadTCGA") and store everything there.

```{r message=FALSE}
library(TCGAbiolinks)

tumor = "OV"
project = paste0("TCGA-", tumor)
genome = "hg38"

methylation_platforms <- c("Illumina Human Methylation 27",
                           "Illumina Human Methylation 450")

dirname = "downloadTCGA"
if (!file.exists(dirname)){
  dir.create(dirname)
}
```

## Clinical data

We can start with the download of clinical data.

```{r message=FALSE, eval=FALSE}
cliQuery <- GDCquery(project = project, data.category = "Clinical", 
                     data.format = "bcr xml")

GDCdownload(cliQuery, method="api", files.per.chunk = 10, 
            directory = "downloadTCGA/GDCdata")

followUp <- GDCprepare_clinic(cliQuery, clinical.info = "follow_up",
                              directory = "downloadTCGA/GDCdata")
newTumorEvent <- GDCprepare_clinic(cliQuery, clinical.info = "new_tumor_event",
                                   directory = "downloadTCGA/GDCdata")
```

We have downloaded two data frames: followUp and newTumorEvent.

## Expression

Now, it's time for the expression counts.

```{r message=FALSE, eval=FALSE}
expQuery <- GDCquery(project = project,
                     data.category = "Transcriptome Profiling",
                     data.type = "Gene Expression Quantification",
                     workflow.type = "STAR - Counts")

GDCdownload(expQuery, method = "api", directory = "downloadTCGA/GDCdata")

exprData <- GDCprepare(expQuery, directory = "downloadTCGA/GDCdata")
```

## Methylation

We need to download methylation beta values, that are available 
from 2 different platforms: "Illumina Human Methylation 27" and "Illumina Human Methylation 450".
Since only 10 samples from the TCGA-OV project have been profiled with the 450k Illumina platform,
we decide to download data from the 27k platform.

```{r message=FALSE, eval=FALSE}
metQuery <- GDCquery(project = project,
                     data.category = "DNA Methylation",
                     data.type = "Methylation Beta Value",
                     platform = methylation_platforms[[1]])

metData <- GDCprepare(metQuery, directory = "downloadTCGA/GDCdata")
```

## Mutation

Next, we download mutation data.

```{r message=FALSE, eval=FALSE}
mutQuery <- GDCquery(
    project = project, 
    data.category = "Simple Nucleotide Variation", 
    data.type = "Masked Somatic Mutation",
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking")

GDCdownload(mutQuery, method = "api", directory = "downloadTCGA/GDCdata")

mutData <- GDCprepare(mutQuery, directory = "downloadTCGA/GDCdata")
```

## Copy Number Variation

Finally, we can download the CNV data generated by the GISTIC2 pipeline.

```{r eval=FALSE}
gisticTable <- getGistic("OV-TP", type = "thresholded")
cnvData <- gisticTable[,-c(1:3)]
colnames(cnvData) <- substr(colnames(cnvData), 1, 12)
row.names(cnvData) <- gisticTable$`Locus ID`
```




We save all the downloaded data in a `.RData` file that we will load in the next tutorial
for the pre-processing.

```{r eval=FALSE, echo=FALSE}
if (!file.exists(paste0(dirname, '/', project, "-", genome, ".RData"))){
    save(exprData, metData, mutData, cnvData, 
         followUp, newTumorEvent, 
         file = paste0(dirname, '/', project, "-", genome, ".RData"), 
         compress = "xz")
}
```




```{r}
sessionInfo()
```

