

```{r message=FALSE}
library(maftools)
library(org.Hs.eg.db)
library(impute)
library(doParallel)
library(MethylMix)
library(TCGAbiolinks)
```


# Pre-processing

```{r message=FALSE}
source("../functions-to-process-TCGA-data.R")

genome = "hg38"
tumor = "OV"
project = paste0("TCGA-", tumor)

# load data downloaded by TCGAbiolinks
load("downloadTCGA/TCGA-OV-hg38.RData")
```

## Expression

```{r}
# convert ensembl ids into entrez
geneAnnotation <- rowData(exprData)

ensembl2entrez <- mapIds(org.Hs.eg.db, 
                         keys = sub("\\..*", "", geneAnnotation$gene_id),
                         column = "ENTREZID", keytype = "ENSEMBL")
geneAnnotation$entrezID <- ensembl2entrez

rowData(exprData) <- geneAnnotation
genes <- geneAnnotation$gene_id[!is.na(geneAnnotation$entrezID)]
```

```{r}
# select primary tumors
colAnnotation <- exprData@colData
primaryTumorSel  <- colAnnotation$definition=="Primary solid Tumor"
primaryTumor <- colnames(exprData)[primaryTumorSel]
```

```{r}
exprData <- exprData[genes, primaryTumor, drop=F]
```

```{r}
# take average for duplicated entrez ids
exp <- assay(exprData, 1)
row.names(exp) <- sub("\\..*", "", row.names(exprData))

avg <- tapply(row.names(exp), rowData(exprData)$entrezID, function(r){
  row = exp[r, ,drop=F]
  if (nrow(row)>1)
    row=colMeans(row, na.rm = T)
  row
})

expAvg <- do.call(rbind, avg)
row.names(expAvg) <- names(avg)

colnames(expAvg) <- substr(colnames(expAvg), 1, 12)

expAvg[1:6, 1:6]
```

## Mutation

```{r}
maf <- read.maf(mutData)

considerThisImpactType <- c("HIGH", "MODERATE")
#considerThisMutationType <- c("Missense_Mutation", "Nonsense_Mutation")

ov.mutations <- prepareMutations(maf@data, impact=considerThisImpactType, 
                                 filterByThisEntrez=row.names(expAvg), 
                                 patients=colnames(expAvg))

ov.mutations$data[1:6,1:6]
```

## CNV

```{r}
ov.cnv <- lapply(cnvData, as.numeric)
ov.cnv <- data.frame(ov.cnv, check.names = F)
row.names(ov.cnv) <- row.names(cnvData)

for (i in names(ov.cnv)) {
  if (!identical(ov.cnv[[i]], as.numeric(cnvData[[i]])))
    warning(paste0("sample number ", i, " is not equal after numeric transformation"))
}

ov.cnv[1:6,1:6]
```

## Methylation

```{r}
# Select primary solid tumors
sampleSelection <- colData(metData27)[[5]] == "Primary solid Tumor"
normalSelection <- colData(metData27)[[5]] == "Solid Tissue Normal"
samples <- colnames(metData27)[sampleSelection]
normals <- colnames(metData27)[normalSelection]
normData <- metData27[, normals]
metData <- metData27[, samples]

# Remove all rows with low values and impute missing values
thr <- ncol(assay(normData))*0.4
remove <- apply(assay(normData),1,sum) > thr
normAssay <- assay(normData)[!remove, ]
normAssay <- imputeKNN(normAssay,k=5)$data

thr <- ncol(assay(metData))*0.4
remove <- apply(is.na(assay(metData)),1,sum) > thr
metAssay <- assay(metData)[!remove, ]
metAssay <- imputeKNN(metAssay,k=5)$data

# make clusters
cl <- makeCluster(16)
registerDoParallel(cl)
metClustValues <- ClusterProbes(metAssay, normAssay)
stopCluster(cl)
```

```{r}
# connect methylation clusters to genes
geneNames <- sapply(strsplit(row.names(metClustValues$MET_Cancer_Clustered), "---"),
                    function(x) x[1])
assList <- tapply(row.names(metClustValues$MET_Cancer_Clustered), geneNames, function(g) g)
symbol2entrez <- mapIds(org.Hs.eg.db, keys=names(assList), column="ENTREZID",
                        keytype="SYMBOL")
names(assList) <- symbol2entrez
metClustValues$eMap <- assList[!is.na(names(assList))]
```

## Survival annotation

```{r}
fup <- createFollowUp(followUp, newTumorEvent)
```

## Two-class annotation

```{r}
subtypes <- PanCancerAtlas_subtypes()
table(subtypes[subtypes$cancer.type=="OVCA", "Subtype_mRNA"])

subtypes <- subtypes[subtypes$cancer.type=="OVCA",]

classes <- data.frame(row.names = substr(subtypes$pan.samplesID, 1, 12), 
                      classes=subtypes$Subtype_mRNA)
```

```{r eval=FALSE}
save(classes, fup, metClustValues, ov.cnv, ov.mutations, expAvg, 
     file="downloadTCGA/TCGA-OV-pre-processed.RData")
```


