---
title: "MOSClip - Survival Analysis"
output:
  html_document:
    toc: true
    toc_float: true
---

# Preface

This tutorial will walk you to perform a complete analysis of multi-omics data associated with survival using MOSClip R package.

MOSClip is a method to combines survival analysis and graphical model theory to test the survival association of pathways or of their connected components that we called modules in a multi-omic framework. Multi-omic gene measurements are tested as covariates of a Cox proportional hazard model after dimensionality reduction of data. The final goal is to find the biological processes impacting the patient’s survival.

MOSClip has a modular structure, allowing the use of one or multiple different omics, as well as different data reduction strategies and tests.

In this tutorial we will focus on the integration of four omics: methylome, transcriptome, genomic mutations and genomic copy number variations, testing if these omics can be sinergically involved in pathways with survival prognostication power.

Furthermore, in MOSClip multiple efforts have been dedicated to the implementation of specific graphical tools to browse, manage and provide help in the interpretation of results. In this tutorial we will also exploit these tools to represent analysis results.

# Preparing the enviornment for the analysis

First we need to load the necessary libraries:
```{r message=FALSE}
# Loading libraries
library(org.Hs.eg.db)
library(EDASeq)
library(MOSClip)
library(graphite)
```
Data retrieving and data analysis will depend on your computational resources. Generally, it needs time and disk space. To speed up this tutorial, we pre-processed the TCGA data for you, which can be downloaded through this link (INSERT HeRE). Details on how to pre-process and format the data are available [here](https://caluralab.github.io/MOSClipTutorials/formatTCGAdatasets.html).

The provided dataset includes matrices, genes per patients, of methylation status, somatic mutations, CNVs, and transcript expression levels of the TCGA ovarian cancer samples.

Move the dataset file inside a directory called "downloadTCGA" in your working directory.
Now, we can load it:

```{r}
# Loading pre-processed data
load("downloadTCGA/TCGA-OV-pre-processed.RData")
```

Next, it is recommended to create a directory to save the analysis results.
```{r}
dirname = "MOSresults"
if (!file.exists(dirname)) { # Checks whether the directory exists
    dir.create(dirname) # If it doesn't, creates a new directory 
}
```
The next step is to modifiy all the multi-omics matrices assigning the type of gene identifier. Here, we will work with Entrez Gene ID. It is indicated with the prefix tag “ENTREZID:”, compliant with Bioconductor org.dbi, as used in the latest graphite R package version. This will allow us to be able to match the omics data to graphite pathways.

```{r}
# Renaming the matrix and then adding the prefix to the gene identifiers
expression <- expAvg
row.names(expression) <- paste0("ENTREZID:", row.names(expression))
mutation <- ov.mutations$data
row.names(mutation) <- paste0("ENTREZID:", row.names(mutation))
names(metClustValues$eMap) <- paste0("ENTREZID:", row.names(metClustValues$eMap))
row.names(ov.cnv) <- paste0("ENTREZID:", row.names(ov.cnv))
```

The TCGA dataset came along with the survival annotations from the table by Liu et al, Cell, 2018. Among the data loaded we can find the object ‘fup’ (short of ‘followup’) that represents our survival annotations. 

We extract the slot for the "progression free survival" (pfs) and we save it in a variable named *survAnnotations*. Then,we want to select only the patients that have samples for all the four omics of our interest. 

```{r}
# Getting survival data 
survAnnotations <- fup$pfs

# Selecting the patients with samples for all omics
survAnnot <- na.omit(survAnnotations)
patients <- row.names(survAnnot)
patients <- intersect(patients, colnames(expression))
patients <- intersect(patients, colnames(metClustValues$MET_Cancer_Clustered))
patients <- intersect(patients, colnames(mutation))
patients <- intersect(patients, colnames(ov.cnv))

survAnnot <- survAnnot[patients, , drop = F]

# If the survAnnot RData file does not exist, create and save it.
if (!file.exists("survAnnot.RData")) {
    file = paste0(dirname, "/survAnnot-", as.character(Sys.Date()), ".RData")
    link = "survAnnot.RData"
    save(survAnnot, file = file)
    file.symlink(file, link)
}

# Selecting the patients for the omics:
methylation <- metClustValues
methylation$MET_Cancer_Clustered <- methylation$MET_Cancer_Clustered[, patients, 
    drop = F]
mutation <- mutation[, patients, drop = F]
cnv <- ov.cnv[, patients, drop = F]
```

If you was enough attentive, you noticed that we didn't perform the patient selection for the expression matrix. That is because we are going to perfom some extra steps after dropping the patients of this matrix. We need to filter the genes to avoid data sparsity, keeping only those genes with at least 100 counts in at least one patients. 

After the selection of patients and genes of the expression matrix, we perform  normalization and log2 of the (counts+1) transformation. This will ensure us to work with expression data that is very similar to a normal distribution, which is the most suitable distribution for the subsequent *MOSClip* tests. The normalization of the data is performed according to the data provided, which can change when the patients and/or samples change. That is why this step is performed after the patient selection and it is present in this tutorial, and not in the previous tutorial ([how to format the dataset for MOSClip](https://caluralab.github.io/MOSClipTutorials/formatTCGAdatasets.html))

```{r}
# Keeping the patients
expression <- expression[, patients, drop = F]

# Filtering the counts 
keep = apply(expression >= 100, 1, any)

# Normalizing the counts
expNorm <- betweenLaneNormalization(expression[keep, , drop = F], which = "upper")
pseudoExpNorm <- log2(expNorm + 1)
```

At this point, we need the a pathway knowledge-base. We will use the Reactome pathway graphs available at *graphite* R package. Reactome is natively distributed in Uniprot, our analysis will be in Entrez gene ID, thus we need to convert the pathway identifiers.

Since we are downloading and converting all the Reactom pathways, this step may take a while. To avoid the boredom of waiting for this step again, we will save the object for future analysis.

```{r}
# If the file doesn't exist, create it
if (file.exists("reactome-entrez.RData")) {
    load("reactome-entrez.RData")
} else {
    reactome <- pathways("hsapiens", "reactome") # getting the Homo sapiens pathways from Reactome
    reactome <- convertIdentifiers(reactome, "entrez") 
    file = paste0(dirname, "/reactome-entrez-", as.character(Sys.Date()), ".RData")
    link = "reactome-entrez.RData"
    save(reactome, file = file)
    file.symlink(file, link)
}
```

The Reactome database has a hierarchical structure, and for this tutorial we are going to analyze only subsets of the pathways. We will cut the pathways according to their sizes. 

For the pathway analysis, we are going to use *reactHuge*, a subset that have all the pathways with more (or equal) than 10 nodes. For module analysis, we are going to use *reactSmall*, so the pathways that are bigger than 20 nodes but smaller that 100 nodes. 

This will ensure a sufficient level of specificity of the pathway/modules and will avoid unusefull reaction redundancies.

```{r}
nodesLength <- sapply(reactome, function(g) {
    length(intersect(graphite::nodes(g), row.names(pseudoExpNorm)))
})
reactSmall <- reactome[nodesLength >= 20 & nodesLength <= 100]
reactHuge <- reactome[nodesLength >= 10]
```

We also need the pathway hierarchy for the *MOSClip* summary plots, so it is convenient to download it now. 
```{r}
pathHierarchy <- downloadPathwayRelationFromReactome()
pathHierarchyGraph <- igraph::graph_from_data_frame(d = pathHierarchy, directed = TRUE)
```

# Creating the Omics object

We are almost ready to run MOSClip! First, we just need to organize the omics matrices in the specific *MOSClip* object: the *Omics* object. This object wraps together all the omics matrices, containing also the survival annotation (or any other type of "colData"), and specific arguments for the dimentionality reduction step. 

Then, we need to indicate the data reduction strategy we want to apply for each matrix/omics. In this tutorial, we chose to use PCA for expression data, cluster analysis for methylation data, vote counting for mutations and CNVs (for detail see [MOSClip paper](https://doi.org/10.1093/nar/gkz324)). This data transformations are easily applied calling *MOSClip* functions, thus here we need only to provide the name of the needed function.

Specifically for the methylation data, *MOSClip* provides the possibility to include a dictionary to associate the methylation level of multiple CpG. This is because it is expected to have more than one CpG cluster associated to a gene. Thus, in the methylation specific arguments you need to provide the dictionary to convert cluster names into genes.

```{r}
multiOmics <- makeOmics(experiments = list(exp = pseudoExpNorm,
                                           met = methylation$MET_Cancer_Clustered,
                                           mut = mutation,
                                           cnv = as.matrix(cnv)),
                        colData = survAnnot,
                        modelInfo = c("summarizeWithPca", "summarizeInCluster",
                                      "summarizeToNumberOfEvents",
                                      "summarizeToNumberOfDirectionalEvents"),
                        specificArgs = list(pcaArgs = list(name = "exp",
                                                           shrink = "FALSE",
                                                           method = "sparse",
                                                           maxPCs = 3),
                                            clusterArgs = list(name = "met",
                                                               dict = methylation$eMap,
                                                               max_cluster_number = 3),
                                            countEvent = list(name = "mut", min_prop = 0.05),
                                            cnvAgv = list(name = "cnv", min_prop = 0.05)))

save(multiOmics, file = "multiOmics.RData")
```

As you can see, in this object we specified the four omic data, we chose four methods for data reduction (one for each omic), and four lists of additional parameters (as described in the help of each reduction function).

# Running MOSClip analysis
*MOSClip* analysis can be performed at the pathway or at the module level, where modules are sub-parts of pathways. A pathway analysis give a general overview of the involved processes, the module analysis can highlights more precisely the mechanism involved.

For both type of analysis, we perform another *a priori* filter, using the genes that are present at least in the expression data, but this is not mandatory.

```{r}
genesToConsider <- row.names(pseudoExpNorm)
```

## Module Analysis
Here we are going to perofm the multi-omics survival analysis on pathway **modules**. 

This step will create a list of *MultiOmicsModules* objects (aka "MOM" objects), in which each of them corresponds to one pathway.

The analysis is quite long, so we will also save the analysis results for future usage. 
```{r warning=FALSE}
if (file.exists("momSurvivalList")) {
    load("momSurvivalList.RData")
} else {
    momSurvivalList <- lapply(reactSmall, function(g) {
        #print(g@title)  # uncomment it so you can see which pathway is being analyzed
        # for each pathway contained in reactSmall, create the MOM object
        res <- multiOmicsSurvivalModuleTest(multiOmics, graph = g,
                                            useTheseGenes = genesToConsider)
        res
    })
    save(momSurvivalList, file = "MOSresults/momSurvivalList.RData")
}
```

## Pathway analysis
In pathway test, the topology of the pathways (in and out gene connections) can be exploited to guide the data reduction step. For this analysis, we suggest to use the topological PCA instead of the sparse PCA, which can be performed by changing the settings in the Omics object.
```{r}
multiOmicsPathway <- multiOmics
multiOmicsPathway@specificArgs$pcaArgs$method = "topological"
multiOmicsPathway@specificArgs$pcaArgs$shrink = TRUE
```

Then we can run the analysis using the function *multiOmicsSurvivalPathwayTest()*:
```{r warning=FALSE}
if (file.exists("mopSurvivalList")) {
    load("mopSurvivalList.RData")
} else {
    mopSurvivalList <- lapply(reactSmall, function(g) {
        #print(g@title)  # uncomment it so you can see which pathway is being analyzed
        #for each pathway contained in reactSmall, create the MOM object
        res <- multiOmicsSurvivalPathwayTest(multiOmicsPathway, graph = g,
                                             useTheseGenes = genesToConsider)
        res
    })
    save(mopSurvivalList, file = "MOSresults/mopSurvivalList.RData")
}
```
This step created a list of *MultiOmicsPathway* objects ("MOP" objects for short), in which each of them corresponds to one pathway. This step also takes a while, so it is better to save the list for future analyses.

Now the analyses are done and we are ready to check it! *MOSClip* has plenty of functions to explore the results. We will show you some examples in the following part.

# Exploring the results
## Results summary
Using the function *multiPathwayModuleReport()* or *multiPathwayReport()* we can plot the tabular summary of the top 10 modules or pathways, selected by p-value of the Cox proportional hazard model.
```{r moduleSummary}
moduleSummary <- multiPathwayModuleReport(momSurvivalList)
```
```{r echo=FALSE}
knitr::kable(moduleSummary[1:10,2:ncol(moduleSumary)])
```

```{r}
pathwaySummary <- multiPathwayReport(mopSurvivalList)
```
```{r echo=FALSE}
knitr::kable(pathwaySummary[1:10,])
```

## Graphical exploration of MOSClip results
MOSClip have a function that can plot a heatmap of the report of the results. The heatmap is sorted by the p-value of the Cox proportional hazard model, using all the omics as covariates. The leftmost column is the p-value of the model, and the other columns are the p-values for each covariate. The color gradient also corresponds to the p-values. In this way, the plot can help to understand the involvment of different omics in the survival. 

### Report as heatmap
For the module analysis, the function is *plotModuleReport()* and it will plot all the modules of a chosen pathway. Please note that this function takes as input a list of MOM objects.
```{r}
plotModuleReport(momSurvivalList[["Activation of Matrix Metalloproteinases"]])
```
As you can see in this heatmap, the module 5 of the pathway "Activation of Matrix Metalloproteinases" (which in plotted in the previous example) is significant. Furthermore, we can infer the omics that drives this survival behavior from the pvalues of the model covariates: expression (expPC1 and PC2), and methylation (met2k2).

For the pathway analysis, the function *plotMultiPathwayReport()* plots the first n pathways in a list of MOP objects:
```{r}
plotMultiPathwayReport(mopSurvivalList, 10)
```
Here we can see that the same pathway we chose before (the "Activation of Matrix Metalloproteinases") also remains signficiant when performing the pathway-type analysis.

### Kaplan-Meier plots
Now that we could identify significant pathway/modules and the covariates that are implicated with survival, we can also plot Kaplan-Meier curves, dividing patients in groups with different omics patterns:
```{r}
plotModuleKM(momSurvivalList[["Activation of Matrix Metalloproteinases"]], 5,
             formula = "Surv(days, status) ~ expPC1 + met2k",
             paletteNames = "Paired", inYears = TRUE)

plotPathwayKM(mopSurvivalList[["Activation of ATR in response to replication stress"]], 
             formula = "Surv(days, status) ~ cnvPOS + expPC1",
             paletteNames = "Paired", inYears = TRUE)
```

### Heatmap 
Finally, we can look at the predictive genes using heatmap and patient additional annotations.
For this step, let's perform some additional formatting of the survival annotation data so we have a better heatmap:
```{r}
additionalA <- survAnnot
additionalA$status[additionalA$status == 1] <- "event"
additionalA$status[additionalA$status == 0] <- "no_event"
additionalA$PFS <- as.factor(additionalA$status)
additionalA$status <- NULL
additionalA$years <- round(additionalA$days/365.24, 0)
additionalA$days <- NULL
```

Then we can finally plot it:
```{r}
plotModuleHeat(momSurvivalList[["Activation of Matrix Metalloproteinases"]], 5,
               sortBy = c("expPC1", "met2k", "status", "days"),
               additionalAnnotations = survAnnot,
               additionalPaletteNames = list(status = "teal", days = "violet"),
               withSampleNames = F)

plotPathwayHeat(mopSurvivalList[["Activation of ATR in response to replication stress"]],
               sortBy = c("cnvPOS", "expPC1", "status", "days"),
               additionalAnnotations = survAnnot,
               additionalPaletteNames = list(status = "teal", days = "violet"),
               withSampleNames = F)
```
When looking to predictive genes, it is extremely useful to correlate them to survival behavior (death event or survival time), as well as other types of annotation (e.g. tumor grade). That’s why MOSClip allows to plot heatmaps with additional custom annotations. 
### Module in graph plot
The function *plotModuleInGraph()* is specific for the module-type analysis, and it allows the visualization of the position of a chosen module inside its pathway. The function takes as input a MOM object, a "*PathwayList*" object from the *graphite* R package, which in our case is the variable *reactSmall*, and the module number we want to visualize:
```{r}
plotModuleInGraph(momSurvivalList[["Activation of Matrix Metalloproteinases"]],
                  reactSmall, moduleNumber = 5)
```

# Resampling strategy
So far, we have identified some modules or pathways that are significant. However, MOSClip gives the possibility to prioritize the most important and stable results with the resampling strategy. In this way, modules or pathways with a p-value <= 0.05 are resampled n times (customized) and later on select those successful above a certain threshold.
In practice, this is done as below:
```{r}
useThesePathwaysM <- unique(moduleSummary$pathway[moduleSummary$pvalue <= 0.05])

if (file.exists("permsModules.RData")) {
    load("permsModules.RData")
} else {
    permsModules <- resamplingModulesSurvival(fullMultiOmics = multiOmics,
                                              reactSmall, nperm = 100,
                                              pathwaySubset = useThesePathwaysM,
                                              genesToConsider = genesToConsider)
    save(permsModules, file = "MOSresults/permsModules.RData")
}
```
In our case we used 100 as number of permutations, and by default this function remove 3 patients per permutation (modifiable). Note that depending on the number of permutation set, this step could take a while to run. 

For the pathway-type analysis is pretty much the same, only that tthe resampling function is different:
```{r}
useThesePathwaysP <- row.names(pathwaySummary[pathwaySummary$pvalue <= 0.05, ])
if (file.exists("permsPathways.RData")) {
    load("permsPathways.RData")
} else {
    permsPathways <- resamplingPathwaySurvival(fullMultiOmics = multiOmicsPathway,
                                               reactHuge, nperm = 100,
                                               pathwaySubset = useThisPathwaysP)
    save(permsPathways, file = "MOSresults/permsPathways.RData")
}
```

## Exploring resampling results
Once the resampling analyses are completed, we can explore the results.
First, we can plot the distribution modules/pathways according to the success count and pvalue:
```{r}
sModule <- moduleSummary[moduleSummary$pathway %in% useThesePathwaysM, , drop = T]
stableModulesSummary <- selectStablePathwaysModules(perms = permsModules,
                                                    moduleSummary = sModule,
                                                    success = 80)

# For the pathway-type results:
sPathway <- pathwaySummary[useThesePathwaysP, , drop = T]
stablePathwaysSummary <- selectStablePathwaysModules(perms = permsPathways,
                                                    moduleSummary = sPathway,
                                                    success = 80)
```



